pkg/webhook/server/server.go

```diff
@@ -104,26 +113,6 @@ func (s *Server) checkTLS() error {
 		return errors.Wrap(err, "failed checking TLS cert file stats")
 	}
 
-	key, err := ioutil.ReadFile(path.Join(s.webhookServer.CertDir, corev1.TLSPrivateKeyKey))
-	if err != nil {
-		return errors.Wrap(err, "failed reading for TLS key")
-	}
-
-	certPEM, err := ioutil.ReadFile(path.Join(s.webhookServer.CertDir, corev1.TLSCertKey))
-	if err != nil {
-		return errors.Wrap(err, "failed reading for TLS cert")
-	}
-
-	caPEM, err := s.certManager.CABundle()
-	if err != nil {
-		return errors.Wrap(err, "failed to retrieve CA cert")
-	}
-
-	err = triple.VerifyTLS(certPEM, key, caPEM)
-	if err != nil {
-		return errors.Wrapf(err, "failed verifying %s/%s", certsPath, keyPath)
-	}
-
```

Comment:
we have to verify the chain here

Reply:
Why?

If anything, it seems arbitrary to do it only just once at start knowing that the certificates might be rotated later on. I don't even know why we need to check for the key/cert files. The underlying server implementation already has a fsnotify watcher doing it.

Reply:
If we start the webhook server with the "empty" placeholder secret to allow pod to start the pod will fail and go to restart mode and that's problematic.

we fixed it by checking TLS before starting the webhook server.

Remember that cert rotation mechanism runs in the same pod as webhook server so we have a chicken egg issue here and we create a placeholder secret with garbage TLS key/cert so pod can start.

Reply:
OK, I am not absolutely convinced that the server should fail & restart missing the TLS key/cert, that is probably worth a revisit.

Anyway, we could call `certManager.VerifyTLS` instead. We probably need to mutex that against `certManager.reconcileCertificates` to be on the safe side. Thoughts?

Reply:
Since we are accessing already the keys instead of using certManager.VerifyTLS let's create a new chain reading the files and call verifyTLS on the chain, so we don't need a mutex.

Reply:
A new chain is already being used, that's not the reason of the mutex. The reason would be non transactional DB/cache.

Reply:
`verifyTLS` is read only, the mutex would just avoid another poll cycle.

Reply:
What do you mean ?

Reply:
`checkTLS` is called with a polling function every 5 secs for 5 min.

Reply:
So additionally to what we talked about @ https://github.com/qinqon/kube-admission-webhook/pull/47#discussion_r610488157, I went for the compromise of adding a flag to manager when verifying so that if active then all provided webhooks must be read successfully and thus cascade on a complete successful verification.

So reconcile is convergent in that not all webhook configs need to exist at once but reconciliation happens for those that do without errors, while verification requires that a valid certificate chain exists for all webhook configs. 

Check `manager.go`& `server.go`